services/webtheme/webtheme.go

```diff
@@ -12,63 +13,156 @@ import (
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/public"
 	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util"
 )
 
 var (
-	availableThemes    []string
-	availableThemesSet container.Set[string]
-	themeOnce          sync.Once
+	availableThemes             []*ThemeMetaInfo
+	availableThemeInternalNames container.Set[string]
+	themeOnce                   sync.Once
 )
 
+const (
+	fileNamePrefix = "theme-"
+	fileNameSuffix = ".css"
+)
+
+type ThemeMetaInfo struct {
+	FileName           string
+	InternalName       string
+	DisplayName        string
+	PreferColorSchemes container.Set[string]
+}
+
+func parseThemeMetaInfoToMap(cssContent string) map[string]string {
+	metaInfoContent := cssContent
+	if pos := strings.LastIndex(metaInfoContent, "gitea-theme-meta-info"); pos >= 0 {
+		metaInfoContent = metaInfoContent[pos:]
+	}
+
+	reMetaInfoItem := `
+(
+\s*(--[-\w]+)
+\s*:
+\s*("(\\"|[^"])*")
+\s*;
+\s*
+)
+`
+	reMetaInfoItem = strings.ReplaceAll(reMetaInfoItem, "\n", "")
+	reMetaInfoBlock := `\bgitea-theme-meta-info\s*\{(` + reMetaInfoItem + `+)\}`
+	re := regexp.MustCompile(reMetaInfoBlock)
+	matchedMetaInfoBlock := re.FindAllStringSubmatch(metaInfoContent, -1)
+	if len(matchedMetaInfoBlock) == 0 {
+		return nil
+	}
+	re = regexp.MustCompile(strings.ReplaceAll(reMetaInfoItem, "\n", ""))
+	matchedItems := re.FindAllStringSubmatch(matchedMetaInfoBlock[0][1], -1)
+	m := map[string]string{}
+	for _, item := range matchedItems {
+		v := item[3]
+		v = strings.TrimPrefix(v, "\"")
+		v = strings.TrimSuffix(v, "\"")
+		v = strings.ReplaceAll(v, `\"`, `"`)
+		m[item[2]] = v
+	}
+	return m
+}
+
+// @media (prefers-color-scheme: dark)
+func parseThemePreferColorSchemes(cssContent string) container.Set[string] {
+	re := regexp.MustCompile(`@media\s*\(\s*prefers-color-scheme\s*:\s*([-\w]+)\s*\)`)
+	matched := re.FindAllStringSubmatch(cssContent, -1)
+	if len(matched) == 0 {
+		return nil
+	}
+	schemes := container.Set[string]{}
+	for _, m := range matched {
+		schemes.Add(m[1])
+	}
+	return schemes
+}
+
+func defaultThemeMetaInfoByFileName(fileName string) *ThemeMetaInfo {
+	themeInfo := &ThemeMetaInfo{
+		FileName:     fileName,
+		InternalName: strings.TrimSuffix(strings.TrimPrefix(fileName, fileNamePrefix), fileNameSuffix),
+	}
+	themeInfo.DisplayName = themeInfo.InternalName
+	return themeInfo
+}
+
+func defaultThemeMetaInfoByInternalName(fileName string) *ThemeMetaInfo {
+	return defaultThemeMetaInfoByFileName(fileNamePrefix + fileName + fileNameSuffix)
+}
+
+func parseThemeMetaInfo(fileName, cssContent string) *ThemeMetaInfo {
+	themeInfo := defaultThemeMetaInfoByFileName(fileName)
+	themeInfo.PreferColorSchemes = parseThemePreferColorSchemes(cssContent)
```

Comment:
What's the purpose of this `prefers-color-scheme` parsing?

Reply:
Could be used to detect "auto" theme in the future. If we decide not using it, it could also be removed easily.